src/libraries/System.Runtime/tests/System/StringTests.cs

```diff
@@ -508,6 +507,68 @@ public static void EnumerateRunes(char[] chars, int[] expected)
+        }
+
+        [Theory]
+        [InlineData("", new[] { "" })]
+        [InlineData("abc", new[] { "abc" })]
+        [InlineData("<CR>", new[] { "", "" })] // empty sequences before and after the CR
+        [InlineData("<CR><CR>", new[] { "", "", "" })] // empty sequences before and after the CR (CR doesn't consume CR)
+        [InlineData("<CR><LF>", new[] { "", "" })] // CR should swallow any LF which follows
+        [InlineData("a<CR><LF><LF>z", new[] { "a", "", "z" })] // CR should swallow only a single LF which follows
+        [InlineData("a<CR>b<LF>c", new[] { "a", "b", "c" })] // CR shouldn't swallow anything other than LF
+        [InlineData("aa<CR>bb<LF><CR>cc", new[] { "aa", "bb", "", "cc" })] // LF shouldn't swallow CR which follows
+        [InlineData("a<CR>b<VT>c<LF>d<NEL>e<FF>f<PS>g<LS>h", new[] { "a", "b<VT>c", "d", "e", "f", "g", "h" })] // VT not recognized as NLF
+        [InlineData("xyz<NEL>", new[] { "xyz", "" })] // sequence at end produces empty string
+        [InlineData("<NEL>xyz", new[] { "", "xyz" })] // sequence at beginning produces empty string
+        [InlineData("abc<NAK>%def", new[] { "abc<NAK>%def" })] // we don't recognize EBCDIC encodings for LF (see Unicode Standard, Sec. 5.8, Table 5-1)
+        public static void ReplaceLineEndings(string input, string[] expectedSegments)
+        {
+            input = FixupSequences(input);
+            expectedSegments = Array.ConvertAll(expectedSegments, FixupSequences);
+
+            // Try Environment.NewLine (and parameterless ctor)
+
+            string expectedEnvNewLineConcat = string.Join(Environment.NewLine, expectedSegments);
+            Assert.Equal(expectedEnvNewLineConcat, input.ReplaceLineEndings());
+            Assert.Equal(expectedEnvNewLineConcat, input.ReplaceLineEndings(Environment.NewLine));
+
+            // Try removing newlines entirely
+
+            Assert.Equal(string.Concat(expectedSegments) /* no joiner */, input.ReplaceLineEndings(""));
+
+            // And try using a custom separator
+
+            Assert.Equal(string.Join("<SEPARATOR>", expectedSegments), input.ReplaceLineEndings("<SEPARATOR>"));
+
+            if (expectedSegments.Length == 1)
+            {
+                // If no newline sequences at all, we should return the original string instance as an optimization
+                Assert.Same(input, input.ReplaceLineEndings());
+                Assert.Same(input, input.ReplaceLineEndings(Environment.NewLine));
+                Assert.Same(input, input.ReplaceLineEndings(""));
+                Assert.Same(input, input.ReplaceLineEndings("<SEPARATOR>"));
+            }
+
+            static string FixupSequences(string input)
+            {
+                // We use <XYZ> markers so that the original strings show up better in the xunit test runner
+
+                return input.Replace("<CR>", "\r")
+                    .Replace("<LF>", "\n")
+                    .Replace("<VT>", "\v")
```

Comment:
> .Replace("<VT>", "\v") [](http://example.com/codeflow?start=19&length=23)

interesting you have the VT test while you didn't include it in the parsing code.

Reply:
It's a negative test to make sure that we _don't_ break on VT. The text of Unicode Sec. 5.8 goes into a little bit of history behind why this is. But the gist of it is that no platform ever standardized on VT as a newline function (NLF).

> Once NLF was reinterpreted to stand for a paragraph separator, in some cases another control character was pressed into service as a line separator. For example, vertical tabulation VT is used in Microsoft Word. However, the choice of character for line separator is even less standardized than the choice of character for NLF.

Per Recommendation R4 in Sec. 5.8, they list the sequences to be searched for as { NLF, LS, PS, FF }. Table 5-2 defines NLF as the set { CR, LF, CRLF, NEL }. So that brings the total set of sequences we're searching for as { CR, LF, CRLF, NEL, LS, PS, FF }. VT is specifically excluded from this list.

Reply:
ok, thanks for the explanation. do you think is worth it to add some comment why VT is not included? I know you have a comment talking about R4 Sec. 5.8 but the comment not mentioning VT.